File: GameShelf/ViewModels/EpicAccountViewModel.swift
//
//  EpicAccountViewModel.swift
//  GameShelf
//

import Foundation
import SwiftData

/// La conexion con la cuenta de Epic: conectar, renovar y desconectar.
///
/// Guarda las credenciales en el Keychain, igual que PSN. Aca importa aun mas:
/// Epic avisa en su propia pagina de que ese codigo **da acceso completo a la
/// cuenta**.
@Observable
@MainActor
final class EpicAccountViewModel {

  /// En que punto esta la conexion.
  enum State: Equatable {
    case desconectado
    case trabajando
    case conectado
    /// Hay credenciales guardadas pero ya no sirven.
    case necesitaCodigoNuevo(message: String, suggestion: String?)
    /// Fallo por otra cosa, por ejemplo la red.
    case fallo(message: String, suggestion: String?)

    var isWorking: Bool { self == .trabajando }
    var isConnected: Bool { self == .conectado }
  }

  private(set) var state: State = .desconectado

  /// Nombre de la cuenta conectada, si Epic lo dio.
  private(set) var displayName: String?

  /// Cuando habra que volver a generar el codigo.
  private(set) var reconnectBy: Date?

  private let service: EpicAuthenticating
  private let keychain: KeychainStoring

  private enum Clave {
    static let accessToken = "epic.accessToken"
    static let refreshToken = "epic.refreshToken"
    static let expiresAt = "epic.expiresAt"
    static let refreshExpiresAt = "epic.refreshExpiresAt"
    static let accountID = "epic.accountID"
    static let displayName = "epic.displayName"
  }

  init(service: EpicAuthenticating = EpicAuthService(), keychain: KeychainStoring = KeychainStore()) {
    self.service = service
    self.keychain = keychain
    cargarEstadoGuardado()
  }

  // MARK: - Conectar

  /// Canjea el codigo que el usuario pego.
  func connect(authorizationCode: String) async {
    guard !state.isWorking else { return }
    state = .trabajando

    do {
      let credenciales = try await service.signIn(authorizationCode: authorizationCode)
      try guardar(credenciales)
      state = .conectado
    } catch {
      state = Self.estado(para: error)
    }
  }

  /// Devuelve un token de acceso vigente, renovandolo si hace falta.
  @discardableResult
  func validAccessToken(at ahora: Date = Date()) async throws -> String {
    guard let guardadas = try credencialesGuardadas() else {
      state = .desconectado
      throw EpicAuthError.sinCredenciales
    }

    guard guardadas.isExpired(at: ahora) else {
      state = .conectado
      return guardadas.accessToken
    }

    do {
      let renovadas = try await service.refresh(using: guardadas.refreshToken)
      try guardar(renovadas)
      state = .conectado
      return renovadas.accessToken
    } catch {
      state = Self.estado(para: error)
      throw error
    }
  }

  /// El identificador de cuenta, que hara falta para pedir la biblioteca.
  func accountID() throws -> String? {
    try keychain.string(for: Clave.accountID)
  }

  /// Borra las credenciales del llavero.
  func disconnect() {
    for clave in [
      Clave.accessToken, Clave.refreshToken, Clave.expiresAt,
      Clave.refreshExpiresAt, Clave.accountID, Clave.displayName
    ] {
      try? keychain.remove(for: clave)
    }
    displayName = nil
    reconnectBy = nil
    state = .desconectado
  }

  // MARK: - Biblioteca

  /// En que punto esta la sincronizacion de la biblioteca de Epic.
  enum LibraryState: Equatable {
    case idle
    case syncing
    case succeeded(created: Int, updated: Int, merged: Int)
    case failed(message: String)

    var isSyncing: Bool { self == .syncing }
  }

  private(set) var libraryState: LibraryState = .idle

  /// Trae los juegos de Epic y los guarda.
  ///
  /// No propaga errores: los deja en `libraryState`. Si la sesion caduco, el
  /// estado de la cuenta pasa a pedir un codigo nuevo.
  func syncLibrary(into context: ModelContext, using client: HTTPClient = URLSessionHTTPClient()) async {
    guard !libraryState.isSyncing else { return }
    libraryState = .syncing

    let servicio = EpicLibraryService(
      client: client,
      accessToken: { [weak self] in
        guard let self else { throw EpicAuthError.sinCredenciales }
        return try await self.validAccessToken()
      },
      accountID: { [weak self] in try self?.accountID() }
    )

    do {
      let juegos = try await servicio.fetchOwnedGames()
      let resultado = try EpicLibrarySyncer.sync(juegos, into: context)
      libraryState = .succeeded(
        created: resultado.created,
        updated: resultado.updated,
        merged: resultado.merged
      )
    } catch let error as EpicAuthError {
      state = Self.estado(para: error)
      libraryState = .failed(message: error.errorDescription ?? "")
    } catch let error as NetworkError {
      libraryState = .failed(
        message: error.errorDescription ?? String(localized: "No se pudo sincronizar.", comment: "Error generico")
      )
    } catch {
      libraryState = .failed(message: error.localizedDescription)
    }
  }

  // MARK: - Guardado

  private func cargarEstadoGuardado() {
    guard let credenciales = (try? credencialesGuardadas()) ?? nil else {
      state = .desconectado
      return
    }

    displayName = credenciales.displayName
    reconnectBy = credenciales.refreshExpiresAt
    state = .conectado
  }

  private func credencialesGuardadas() throws -> EpicCredentials? {
    guard let access = try keychain.string(for: Clave.accessToken),
          let refresh = try keychain.string(for: Clave.refreshToken),
          let vence = try keychain.string(for: Clave.expiresAt),
          let segundos = TimeInterval(vence)
    else { return nil }

    let refrescoVence = (try? keychain.string(for: Clave.refreshExpiresAt))
      .flatMap { $0 }
      .flatMap(TimeInterval.init)
      .map(Date.init(timeIntervalSince1970:))

    return EpicCredentials(
      accessToken: access,
      refreshToken: refresh,
      expiresAt: Date(timeIntervalSince1970: segundos),
      refreshExpiresAt: refrescoVence,
      accountID: try keychain.string(for: Clave.accountID),
      displayName: try keychain.string(for: Clave.displayName)
    )
  }

  private func guardar(_ credenciales: EpicCredentials) throws {
    try keychain.set(credenciales.accessToken, for: Clave.accessToken)
    try keychain.set(credenciales.refreshToken, for: Clave.refreshToken)
    try keychain.set(String(credenciales.expiresAt.timeIntervalSince1970), for: Clave.expiresAt)

    // Estos tres solo se escriben si vinieron: una renovacion no siempre los
    // repite, y borrarlos dejaria la pantalla sin datos que ya se conocian.
    if let vence = credenciales.refreshExpiresAt {
      try keychain.set(String(vence.timeIntervalSince1970), for: Clave.refreshExpiresAt)
      reconnectBy = vence
    }
    if let cuenta = credenciales.accountID {
      try keychain.set(cuenta, for: Clave.accountID)
    }
    if let nombre = credenciales.displayName {
      try keychain.set(nombre, for: Clave.displayName)
      displayName = nombre
    }
  }

  /// Traduce un error al estado que corresponde.
  private static func estado(para error: Error) -> State {
    if let epic = error as? EpicAuthError {
      let mensaje = epic.errorDescription ?? ""
      return epic.necesitaCodigoNuevo
        ? .necesitaCodigoNuevo(message: mensaje, suggestion: epic.recoverySuggestion)
        : .fallo(message: mensaje, suggestion: epic.recoverySuggestion)
    }

    if let red = error as? NetworkError {
      return .fallo(message: red.errorDescription ?? "", suggestion: red.recoverySuggestion)
    }

    return .fallo(message: error.localizedDescription, suggestion: nil)
  }
}

File: android/app/src/main/java/com/gameshelf/ui/accounts/EpicAccountViewModel.kt
package com.gameshelf.ui.accounts

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.gameshelf.data.epic.EpicAuthError
import com.gameshelf.data.epic.EpicAuthService
import com.gameshelf.data.epic.EpicAuthenticating
import com.gameshelf.data.epic.EpicCredentials
import com.gameshelf.data.epic.EpicLibraryService
import com.gameshelf.data.net.HttpClient
import com.gameshelf.data.net.OkHttpNetworkClient
import com.gameshelf.data.repository.GameStore
import com.gameshelf.data.secrets.SecureStoring
import com.gameshelf.data.sync.EpicLibrarySyncer
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import java.time.Instant

/**
 * La conexion con la cuenta de Epic: conectar, renovar y desconectar.
 *
 * Es el gemelo de [PSNAccountViewModel] y guarda las credenciales igual de
 * cifradas, por la misma razon: el codigo de Epic da acceso completo a la
 * cuenta, y su propia pagina lo advierte.
 */
class EpicAccountViewModel(
  private val service: EpicAuthenticating = EpicAuthService(),
  private val secure: SecureStoring,
  private val store: GameStore,
  private val client: HttpClient = OkHttpNetworkClient(),
) : ViewModel() {

  /** En que punto esta la conexion. */
  sealed interface State {
    data object Desconectado : State
    data object Trabajando : State
    data object Conectado : State

    /** Hay credenciales guardadas pero ya no sirven. */
    data class NecesitaCodigoNuevo(val error: Throwable) : State

    /** Fallo por otra cosa, por ejemplo la red. */
    data class Fallo(val error: Throwable) : State

    val isWorking: Boolean get() = this is Trabajando
    val isConnected: Boolean get() = this is Conectado
  }

  private val _state = MutableStateFlow<State>(State.Desconectado)
  val state: StateFlow<State> = _state.asStateFlow()

  private val _expiresAt = MutableStateFlow<Instant?>(null)
  val expiresAt: StateFlow<Instant?> = _expiresAt.asStateFlow()

  /** Cuando habra que volver a generar el codigo. Epic si manda este dato. */
  private val _reconnectBy = MutableStateFlow<Instant?>(null)
  val reconnectBy: StateFlow<Instant?> = _reconnectBy.asStateFlow()

  /** Nombre visible de la cuenta conectada. Solo para mostrarlo. */
  private val _displayName = MutableStateFlow<String?>(null)
  val displayName: StateFlow<String?> = _displayName.asStateFlow()

  init {
    cargarEstadoGuardado()
  }

  // --- Conectar ------------------------------------------------------------

  fun connect(authorizationCode: String) {
    if (_state.value.isWorking) return
    _state.value = State.Trabajando

    viewModelScope.launch {
      try {
        guardar(service.signIn(authorizationCode))
        _state.value = State.Conectado
      } catch (e: Throwable) {
        _state.value = estado(e)
      }
    }
  }

  /**
   * Devuelve un token de acceso vigente, renovandolo si hace falta.
   *
   * @throws EpicAuthError si no hay credenciales o si ya no se pueden renovar.
   */
  suspend fun validAccessToken(ahora: Instant = Instant.now()): String {
    val guardadas = credencialesGuardadas() ?: run {
      _state.value = State.Desconectado
      throw EpicAuthError.SinCredenciales
    }

    if (!guardadas.isExpired(ahora)) {
      _state.value = State.Conectado
      return guardadas.accessToken
    }

    try {
      val renovadas = service.refresh(guardadas.refreshToken)
      guardar(renovadas)
      _state.value = State.Conectado
      return renovadas.accessToken
    } catch (e: Throwable) {
      _state.value = estado(e)
      throw e
    }
  }

  fun disconnect() {
    listOf(ACCESS_TOKEN, REFRESH_TOKEN, EXPIRES_AT, REFRESH_EXPIRES_AT, ACCOUNT_ID, DISPLAY_NAME)
      .forEach(secure::remove)
    _expiresAt.value = null
    _reconnectBy.value = null
    _displayName.value = null
    _state.value = State.Desconectado
  }

  // --- Biblioteca ----------------------------------------------------------

  sealed interface LibraryState {
    data object Idle : LibraryState
    data object Syncing : LibraryState
    data class Succeeded(val created: Int, val updated: Int, val merged: Int) : LibraryState
    data class Failed(val error: Throwable) : LibraryState

    val isSyncing: Boolean get() = this is Syncing
  }

  private val _libraryState = MutableStateFlow<LibraryState>(LibraryState.Idle)
  val libraryState: StateFlow<LibraryState> = _libraryState.asStateFlow()

  /** Trae los juegos de Epic y los guarda. No propaga errores. */
  fun syncLibrary() {
    if (_libraryState.value.isSyncing) return
    _libraryState.value = LibraryState.Syncing

    val servicio = EpicLibraryService(
      client = client,
      accessToken = { validAccessToken() },
      accountID = { secure.string(ACCOUNT_ID) },
    )

    viewModelScope.launch {
      try {
        val juegos = servicio.fetchOwnedGames()
        val resultado = EpicLibrarySyncer.sync(juegos, store)
        _libraryState.value = LibraryState.Succeeded(
          resultado.created, resultado.updated, resultado.merged,
        )
      } catch (e: Throwable) {
        if (e is EpicAuthError) _state.value = estado(e)
        _libraryState.value = LibraryState.Failed(e)
      }
    }
  }

  // --- Guardado ------------------------------------------------------------

  private fun cargarEstadoGuardado() {
    val credenciales = runCatching { credencialesGuardadas() }.getOrNull() ?: run {
      _state.value = State.Desconectado
      return
    }

    _expiresAt.value = credenciales.expiresAt
    _reconnectBy.value = credenciales.refreshExpiresAt
    _displayName.value = credenciales.displayName
    _state.value = State.Conectado
  }

  private fun credencialesGuardadas(): EpicCredentials? {
    val access = secure.string(ACCESS_TOKEN) ?: return null
    val refresh = secure.string(REFRESH_TOKEN) ?: return null
    val vence = secure.string(EXPIRES_AT)?.toLongOrNull() ?: return null

    return EpicCredentials(
      accessToken = access,
      refreshToken = refresh,
      expiresAt = Instant.ofEpochMilli(vence),
      refreshExpiresAt = secure.string(REFRESH_EXPIRES_AT)?.toLongOrNull()
        ?.let(Instant::ofEpochMilli),
      accountID = secure.string(ACCOUNT_ID),
      displayName = secure.string(DISPLAY_NAME),
    )
  }

  private fun guardar(credenciales: EpicCredentials) {
    secure.set(credenciales.accessToken, ACCESS_TOKEN)
    secure.set(credenciales.refreshToken, REFRESH_TOKEN)
    secure.set(credenciales.expiresAt.toEpochMilli().toString(), EXPIRES_AT)
    _expiresAt.value = credenciales.expiresAt

    credenciales.refreshExpiresAt?.let { vence ->
      secure.set(vence.toEpochMilli().toString(), REFRESH_EXPIRES_AT)
      _reconnectBy.value = vence
    }

    // El id de cuenta hace falta para pedir los tiempos jugados, y el refresco
    // no siempre lo repite: se conserva el que ya se conocia.
    credenciales.accountID?.let { secure.set(it, ACCOUNT_ID) }
    credenciales.displayName?.let {
      secure.set(it, DISPLAY_NAME)
      _displayName.value = it
    }
  }

  private fun estado(error: Throwable): State = when {
    error is EpicAuthError && error.necesitaCodigoNuevo -> State.NecesitaCodigoNuevo(error)
    else -> State.Fallo(error)
  }

  private companion object {
    const val ACCESS_TOKEN = "epic.accessToken"
    const val REFRESH_TOKEN = "epic.refreshToken"
    const val EXPIRES_AT = "epic.expiresAt"
    const val REFRESH_EXPIRES_AT = "epic.refreshExpiresAt"
    const val ACCOUNT_ID = "epic.accountID"
    const val DISPLAY_NAME = "epic.displayName"
  }
}
